Working implementation of Stack, Queue, UnorderedList & BinarySearchTree

Greenheart opened the pull request:
- Linted
- Including tests

However, the coverage is at 81% even though I've got tests for all methods. What's the issue?

![test-coverage](https://cloud.githubusercontent.com/assets/6125097/12465849/9507669e-bfd0-11e5-879c-e1480b9d6cb6.png)

Greenheart committed:
Implemented UnorderedList

Linting passing and most tests included though not with full
coverage.

Greenheart committed:
Implemented a Stack

Tests included but oops, no linting this time!

Greenheart committed:
Implemented a Queue

Including tests, ut not linting this time either.

Linted last commit at least, so close enough! :P

Greenheart committed:
Minor tweaks to tests

I don't understand what the coverage report means exactly, because
even though I've made tests for every method I only get 81% coverage

Greenheart commented:
The build error might be because of the most recent change to README.rst

Limpan commented:
Click on `exercises/list.py` and you'll get the source code with all lines marked appropriately. You should be able to clearly see what line's are left out by your tests.
![image](https://cloud.githubusercontent.com/assets/1254504/12480585/b7dfcf12-c043-11e5-9494-3dede87d5f1d.png)

Greenheart committed:
Fixed bugs in UnorderedList.pop()

The major problem was that I forgot that 0 is a falsey value that
therefore did not pass the checks. Solved this by first checking
that the argument *position* is has a proper index and then that
it is an integer.

Also improved tests to get 98% code-coverage. To ease debugging and
make the UnorderedList-class useful, I also added a helper-method
that returns a copy of all present values in the list.

Greenheart commented:
Linting errors due to unused Exceptions. Where we supposed to throw these instead of for example, returning `False` when a method failed?

Limpan commented:
Yep, at least for some of them. Trying to remove from an empty list is one example where raising `EmptyList` would be appropriate. Searching for an item and failing should still return `False`.

Greenheart committed:
Many changes and hopefully some improvements.

Still a lot of code for simple tasks, but at least it works properly

Greenheart committed:
Improved tests, coverage and no linting-errors

There's just one line gives an error and therefore I've just got 99% coverage.

Greenheart commented:
99% coverage. I don't care about the last error since the function is working properly

Greenheart committed:
Merge branch 'master' of https://github.com/teknik-eksjo/datastrukturer

Conflicts:
	exercises/list.py
	exercises/simple.py

Greenheart committed:
Implemented insert() and lookup()

Greenheart committed:
Implemented delete() + it's tests

The delete-method of BST now seems to work pretty well.
Not the most elegant solution but it works as expected
according to the tests.

Also improved the insert-method to make sure that keys
of nodes are unique.

Greenheart committed:
Implement BST traverse() and __str__() methods.

Includes tests with good coverage and linted code.

Also abstracted away the creation of BSTs with a helper method
that will create new trees from a given sequence of keys.
Unfortunately it does not accept dictionaries with key:value pairs
but for the simple testing done so far in the project, it's enough
if the helper can create trees with just keys.

Greenheart commented:
This PR now contains an implementation of the BST, complete with tests and linted code.

94% coverage due to unused +not tested graph.py and some lines that the coverage-script simply doesn't like. For example a function that ends inside a loop will cause a coverage warning (yellow) because the function never exits the function in the "right way".
